drivers/regulator/regulator_npm1300.c

```diff
@@ -0,0 +1,613 @@
+	int32_t retention_uv;
+	struct gpio_dt_spec enable_gpios;
+	struct gpio_dt_spec retention_gpios;
+	struct gpio_dt_spec pwm_gpios;
+};
+
+struct regulator_npm1300_data {
+	struct regulator_common_data data;
+};
+
+/* Linear range for output voltage, common for all bucks and LDOs on this device */
+static const struct linear_range buckldo_range = LINEAR_RANGE_INIT(1000000, 100000, 0U, 23U);
+
+/* Read multiple registers from specified address */
+static int reg_read_burst(const struct device *dev, uint8_t base, uint8_t offset, uint8_t *data,
+			  size_t len)
+{
+	const struct regulator_npm1300_config *config = dev->config;
+	const struct regulator_npm1300_pconfig *pconfig = config->p->config;
+	uint8_t buff[] = {base, offset};
+
+	return i2c_write_read_dt(&pconfig->i2c, buff, sizeof(buff), data, len);
+}
+
+static int reg_read(const struct device *dev, uint8_t base, uint8_t offset, uint8_t *data)
+{
+	return reg_read_burst(dev, base, offset, data, 1U);
+}
+
+/* Write single register to specified address */
+static int reg_write(const struct device *dev, uint8_t base, uint8_t offset, uint8_t data)
+{
+	const struct regulator_npm1300_config *config = dev->config;
+	const struct regulator_npm1300_pconfig *pconfig = config->p->config;
+	uint8_t buff[] = {base, offset, data};
+
+	return i2c_write_dt(&pconfig->i2c, buff, sizeof(buff));
+}
+
+/* Modify bits in single register */
+static int reg_update(const struct device *dev, uint8_t base, uint8_t offset, uint8_t data,
+		      uint8_t mask)
+{
+	uint8_t reg;
+	int ret;
+
+	ret = reg_read(dev, base, offset, &reg);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	reg = (reg & ~mask) | (data & mask);
+
+	return reg_write(dev, base, offset, reg);
+}
+
+unsigned int regulator_npm1300_count_voltages(const struct device *dev)
+{
+	const struct regulator_npm1300_config *config = dev->config;
+
+	switch (config->source) {
+	case NPM1300_SOURCE_BUCK1:
+	case NPM1300_SOURCE_BUCK2:
+	case NPM1300_SOURCE_LDO1:
+	case NPM1300_SOURCE_LDO2:
+		return linear_range_values_count(&buckldo_range);
+	default:
+		return -ENODEV;
+	}
+}
+
+int regulator_npm1300_list_voltage(const struct device *dev, unsigned int idx, int32_t *volt_uv)
+{
+	const struct regulator_npm1300_config *config = dev->config;
+
+	switch (config->source) {
+	case NPM1300_SOURCE_BUCK1:
+	case NPM1300_SOURCE_BUCK2:
+	case NPM1300_SOURCE_LDO1:
+	case NPM1300_SOURCE_LDO2:
+		return linear_range_get_value(&buckldo_range, idx, volt_uv);
+	default:
+		return -ENODEV;
+	}
+}
+
+static int regulator_npm1300_set_ret_voltage(const struct device *dev, int32_t retention_uv)
+{
+	const struct regulator_npm1300_config *config = dev->config;
+	uint16_t idx;
+	uint8_t chan;
+	int ret;
+
+	switch (config->source) {
+	case NPM1300_SOURCE_BUCK1:
+		chan = 0U;
+		break;
+	case NPM1300_SOURCE_BUCK2:
+		chan = 1U;
+		break;
+	default:
+		return -ENOTSUP;
+	}
+
+	ret = linear_range_get_win_index(&buckldo_range, retention_uv, retention_uv, &idx);
+
+	if (ret == -EINVAL) {
```

Comment:
maybe just `ret < 0`?

Reply:
linear_range_get_win_index returns -ERANGE if the requested window partially overlaps the permitted range.  So -EINVAL is a failure and -ERANGE is a success and they need to be handled separately.

I've been caught out by this before in an earlier PR, should I add a comment to explain the reason for the unexpectedly specific condition?

Reply:
I see, sorry for missing that. It should be ok then.